src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -0,0 +1,1024 @@
+_CONFIG_FOR_DOC = "BLOOMConfig"
+_TOKENIZER_FOR_DOC = "BLOOMTokenizer"
+
+BLOOM_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "bigscience/bigscience-small-testing",
+    # "bigscience/BLOOM",
+    # See all BLOOM models at https://huggingface.co/models?filter=bloom
+]
+
+# Utility functions below:
+
+
+def attention_mask_func(attention_scores, attention_mask, causal_mask):
+    if attention_mask.dtype == torch.bool:
+        attention_mask_bool = ~attention_mask
+    else:
+        attention_mask_bool = (1 - attention_mask).bool()
+
+    query_length, key_length, n_heads = attention_scores.size(2), attention_scores.size(3), attention_scores.size(1)
+    padded_causal_mask = (
+        attention_mask_bool[:, None, key_length - query_length : key_length, None]
+        + ~causal_mask[:, :, key_length - query_length : key_length, :key_length]
+    ).bool()
+    # Make use of floats
+    if padded_causal_mask.dtype == torch.bool:
+        return attention_scores.masked_fill_(padded_causal_mask.expand(-1, n_heads, -1, -1), -10000.0)
+    else:
+        values_to_attend = 1.0 - padded_causal_mask
+        return (values_to_attend * attention_scores) - 10000.0 * padded_causal_mask
+
+
+# def attention_mask_func(attention_scores, attention_mask):
+#     return attention_scores.masked_fill_(attention_mask, -10000.0)
+
+
+def bias_dropout_add(x, bias, residual, prob, training):
+    # type: (Tensor, Tensor, Tensor, float, bool) -> Tensor
+    out = torch.nn.functional.dropout(x + bias, p=prob, training=training)
+    out = residual + out
+    return out
+
+
+def get_bias_dropout_add(training):
+    def _bias_dropout_add(x, bias, residual, prob):
+        return bias_dropout_add(x, bias, residual, prob, training)
+
+    return _bias_dropout_add
+
+
+@torch.jit.script
+def bias_dropout_add_fused_train(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, True)
+
+
+@torch.jit.script
+def bias_dropout_add_fused_inference(x, bias, residual, prob):
+    # type: (Tensor, Tensor, Tensor, float) -> Tensor
+    return bias_dropout_add(x, bias, residual, prob, False)
+
+
+class BLOOMAttention(nn.Module):
+    def __init__(self, config, layer_number=None):
+        super().__init__()
+
+        dtype = getattr(torch, config.dtype)
+        self.pretraining_tp = config.pretraining_tp
+
+        self.hidden_size = config.hidden_size
+        self.num_heads = config.n_head
+        self.head_dim = self.hidden_size // self.num_heads
+        self.split_size = self.hidden_size
+        self.attention_softmax_in_fp32 = config.attention_softmax_in_fp32
+        self.masked_softmax_fusion = config.masked_softmax_fusion
+
+        if dtype == torch.float16:
+            self.fp16 = True
+            self.bf16 = False
+        elif dtype == torch.bfloat16:
+            self.fp16 = False
+            self.bf16 = True
+        else:
+            self.fp16 = False
+            self.bf16 = False
```

Comment:
This could change between the init and the first forward pass. It should probably be updated there.